### ISSUE TITLE
Some containers break when using gRPC FUSE

### ISSUE THREAD

**Issue by gbataille** [docker/for-mac#4964]
- [x] I have tried with the latest version of my channel (Stable or Edge)
  - [x] I have uploaded Diagnostics
  - Diagnostics ID: `AD7CB9C6-4F63-4672-811B-998D45C9DC7B/20201002071449`

I have seen a couple of opened issues around gRPC FUSE but I could not find one that I was sure matched what I'm seeing.

### Expected behavior

With default setup (which now has gRPC FUSE active), I expect standard images (minio, rabbitmq in this case) to work in "standard" configuration, i.e. with mapped volumes to my mac disk to store the data of those application/containers

### Actual behavior

I saw 2 issues so far, with `minio` and `rabbitmq`.  
**Disabling gRPC FUSE solves both issues**

- with `minio/minio:RELEASE.2019-10-11T00-38-09Z`
  Running the container, I can start minio if I use a container path. But if I use a path that is a volume on my host, it fails
  ```bash
  docker run -v $PWD/data/data:/data minio/minio:RELEASE.2019-10-11T00-38-09Z server /data
  ```
  ```
  API: SYSTEM()
  Time: 07:02:00 UTC 10/02/2020
  DeploymentID: 8fb098ed-09ce-4e50-a713-0f431e16b52c
  Error: Unexpected error, please report this issue at https://github.com/minio/minio/issues
      10: cmd/fs-v1-helpers.go:333:cmd.fsCreateFile()
       9: cmd/fs-v1.go:905:cmd.(*FSObjects).putObject()
       8: cmd/fs-v1.go:819:cmd.(*FSObjects).PutObject()
       7: cmd/config-common.go:62:cmd.saveConfig()
       6: cmd/config.go:68:cmd.saveServerConfig()
       5: cmd/config-current.go:514:cmd.newSrvConfig()
       4: cmd/config-migrate.go:2474:cmd.migrateConfigToMinioSys()
       3: cmd/config.go:181:cmd.initConfig()
       2: cmd/config.go:144:cmd.(*ConfigSys).Init()
       1: cmd/server-main.go:342:cmd.serverMain()
  ERROR Unable to initialize config system: Unexpected error, please report this issue at 
  https://github.com/minio/minio/issues
  ```
  gives an error where basically minio can't perform a file system operation. This actually works with `minio/minio:latest` pointing to the fact that it might be related to an "old" file system library.
- with `rabbitmq:3.7.14-management-alpine`
  Running the container once works. Shutting it down there restarting it with the same volume mapping (and therefore existing files) fails
  ```shell
  docker run -v $PWD/data/rabbitmq:/var/lib/rabbitmq rabbitmq:3.7.14-management-alpine
  ```
  ```
  chown: /var/lib/rabbitmq/.erlang.cookie: I/O error
  ```
  This is actually the case also with the latest image `rabbitmq:management`

### Information

  - macOS Version: 10.15.6

### Diagnostic logs

```
Docker for Mac: version...

```

#### with minio

  1. Make sure gRPC FUSE is activated
  2. Start minio with a volume mapping that has never been used by minio (otherwise the config files are already created, minio does not try to write and you don't see the issue)
  ```shell
  docker run -v $PWD/data:/data minio/minio:RELEASE.2019-10-11T00-38-09Z server /data
  ```

#### with rabbitmq

  1. Make sure gRPC FUSE is activated
  2. Start rabbitmq with a volume mapping that has never been used (empty folder)
  ```shell
  docker run -v $PWD/data/rabbitmq:/var/lib/rabbitmq rabbitmq:management
  ```
  3. Stop the container (`Ctrl+C`)
  4. Restart a container with the same volume mapping
  ```shell
  docker run -v $PWD/data/rabbitmq:/var/lib/rabbitmq rabbitmq:management
  ```

**Comment by stephen-turner** [comment]
Thanks, @djs55 is investigating

**Comment by djs55** [comment]
Thanks @gbataille for the clear repro steps. I can reproduce both of these problems:
- `minio` performs an `fallocate(FALLOC_FL_KEEP_SIZE)` but unfortunately we haven't implemented `FALLOC_FL_KEEP_SIZE`
- `rabbitmq` performs the equivalent of `touch foo; chmod 400 foo; chown 100 foo` and unfortunately the attempt to persist the ownership information is blocked by the file permissions.

I've created tickets for these internally.

**Comment by hughsw** [comment]
FYI, I'm seeing a similar problem with MariaDB running in a container using DB files mapped from the Mac host.  With gRPC FUSE enabled, the DB server is reports errors trying to read table data, even though it sees and lists the tables.  If I switch back to osxfs, everything works the way it used to.  I will probably post a separate issue once I've diagnosed a little further.

**Comment by greatvovan** [comment]
One more not working image: `tarantool/tarantool`:

```
docker run \
  --name mytarantool \
  -d -p 3301:3301 \
  -v /data/dir/on/host:/var/lib/tarantool \
  tarantool/tarantool:2.6.0
```

As long as it complains

> can't allocate disk space: Invalid argument

I may assume it has the same problem as minio.

Details: https://stackoverflow.com/questions/64178499/tarantool-does-not-start-due-to-disk-write-error

**Comment by gbataille** [comment]
Awesome, thanks @djs55 

Do you think the default should be changed back? quite frankly it's not easy to track those things at all when they happen (I assumed for a while that we were building our images badly, or getting a corrupted image or something). And it seems there are at least a couple of bugs on reasonably standard file system operations.

I actually warned all my friends so they would not lose time on it

**Comment by djs55** [comment]
A quick status update: there's a dev build (based on 2.4.0.0 stable, a pre-release version of a possible 2.4.0.1 update) https://desktop-stage.docker.com/mac/stable/48820/Docker.dmg with some filesystem fixes:
- `FALLOC_FL_KEEP_SIZE` now returns the expected error code (`EOPNOTSUPP` rather than `EINVAL`)
- we now use Mac filesystem ACLs to ensure we can read the extended attributes containing the ownership info

If you have a chance to try this build, please let me know how it goes. If it fails, another repro example would be very helpful!

Thanks for the clear repros provided so far.

## With minio

```
Dave-MBP:minio djs$ rm -rf data/
Dave-MBP:minio djs$ docker run -v $PWD/data:/data minio/minio:RELEASE.2019-10-11T00-38-09Z server /data
Unable to find image 'minio/minio:RELEASE.2019-10-11T00-38-09Z' locally
RELEASE.2019-10-11T00-38-09Z: Pulling from minio/minio
e7c96db7181b: Pull complete 
d5e5170863b7: Pull complete 
d98aad23d939: Pull complete 
Digest: sha256:0d02f16a1662653f9b961211b21ed7de04bf04492f44c2b7594bacbfcc519eb5
Status: Downloaded newer image for minio/minio:RELEASE.2019-10-11T00-38-09Z

 You are running an older version of MinIO released 1 year ago 
 Update: docker pull minio/minio:RELEASE.2020-10-09T22-55-05Z 


Endpoint:  http://172.17.0.2:9000  http://127.0.0.1:9000

Browser Access:
   http://172.17.0.2:9000  http://127.0.0.1:9000

Object API (Amazon S3 compatible):
   Go:         https://docs.min.io/docs/golang-client-quickstart-guide
   Java:       https://docs.min.io/docs/java-client-quickstart-guide
   Python:     https://docs.min.io/docs/python-client-quickstart-guide
   JavaScript: https://docs.min.io/docs/javascript-client-quickstart-guide
   .NET:       https://docs.min.io/docs/dotnet-client-quickstart-guide
```

## With rabbitmq

```
Dave-MBP:rabbitmq djs$ rm -rf data
Dave-MBP:rabbitmq djs$ docker run -v $PWD/data/rabbitmq:/var/lib/rabbitmq rabbitmq:management
...
2020-10-12 16:44:23.034 [info] <0.636.0> Message store for directory '/var/lib/rabbitmq/mnesia/rabbit@ab05f296e7c4/msg_stores/vhosts/628WB79CIFDYO9LJI6DKMI09L/msg_store_transient' is stopped
Dave-MBP:rabbitmq djs$ 
Dave-MBP:rabbitmq djs$ docker run -v $PWD/data/rabbitmq:/var/lib/rabbitmq rabbitmq:management
Configuring logger redirection
...
2020-10-12 16:44:46.734 [info] <0.681.0> Server startup complete; 4 plugins started.
 * rabbitmq_prometheus
 * rabbitmq_management
 * rabbitmq_web_dispatch
 * rabbitmq_management_agent
2020-10-12 16:44:46.734 [info] <0.681.0> Resetting node maintenance status
```

## With tarantool

```
Dave-MBP:tarantool djs$ docker run -p 3301:3301   -v $PWD/data:/var/lib/tarantool   tarantool/tarantool:2.6.0
...
2020-10-12 16:48:12.505 [1] main/116/console/unix/:/var/run/tarantool/tarantool.sock I> started
2020-10-12 16:48:12.505 [1] main C> entering the event loop
```

**Comment by gbataille** [comment]
good news. thanks @djs55

**Comment by simonferquel** [comment]
The issue has been fixed in Edge release 2.4.2.0, thank you for reporting it.

**Comment by greatvovan** [comment]
I confirm, it works with gRPC FUSE in version 2.4.2.0. Thanks for the work!

**Comment by greatvovan** [comment]
BTW, in this version I can see a UI bug related to this:

![image](https://user-images.githubusercontent.com/4903007/96938313-deb87f00-147e-11eb-8509-2798d554905e.png)

The description of the option is identical to the one for gRPC FUSE. I believe it was a copypaste mistake.

**Comment by punktilious** [comment]
Also affects ibmcom/db2.

**Comment by stephen-turner** [comment]
This is fixed in 2.5.0.0. There were a couple of different causes, so if you do find any other containers that don't work, please open a new ticket.
